Set default post type in class-wp-query.php

jonnynews opened the pull request:
Post type handling in the WP_Query class has been improved. Now, if a post type is not defined, a default one will be set based on whether the post is an attachment, a page, or a regular post. This adjustment streamlines the code, improving both efficiency and readability.

<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/59442

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

jonnynews committed:
Set default post type in class-wp-query.php

Post type handling in the WP_Query class has been improved. Now, if a post type is not defined, a default one will be set based on whether the post is an attachment, a page, or a regular post. This adjustment streamlines the code, improving both efficiency and readability.

jonnynews committed:
Normalize post type handling in WP_Query and update tests

The post type handling in WP_Query has been refined to automatically set a default value when no specific post type is given. This change simplifies the defined arguments and enhances the code's efficiency. Additionally, test cases have been expanded to confirm the correct cache key generation related to the implemented changes.

jonnynews committed:
Update SQL query formatting and refactor cache key test

Corrected the indentation in SQL query to improve code readability. The "test_generate_cache_key_normalize" method has been refactored to handle the fields and request variables accurately, ensuring proper testing of cache key generation across different queries and post types.

jonnynews committed:
Fix lints.

joemcgill commented on src/wp-includes/class-wp-query.php:
```diff
@@ -3106,12 +3106,12 @@ public function get_posts() {
+					 WHERE 1=1 $where
+					 $groupby
+					 $orderby
+					 $limits";
```
I don't think these changes are needed.

jonnynews replied:
Fixed.

joemcgill commented on src/wp-includes/class-wp-query.php:
```diff
@@ -4856,6 +4856,21 @@ protected function generate_cache_key( array $args, $sql ) {
+		}
+		$args['post_type'] = (array) $args['post_type'];
+
+		if ( ! isset( $q['orderby'] ) ) {
+			$args['orderby'] = 'date';
+		}
```
This looks right, based on [this block](https://github.com/WordPress/wordpress-develop/blob/3b9aefb06cae1f34ed19b08caf3b3ed32df04b73/src/wp-includes/class-wp-query.php#L2432-L2442). An inline comment would be helpful here.

It makes me wonder if we should actually remove the `orderby` and just use the ORDER BY clause from the `$sql` here instead, as there are other places where differing `$q['orderby']` params get converted into the same query. I don't see that as a blocker, however.

jonnynews replied:
Added.

joemcgill commented on tests/phpunit/tests/query/cacheResults.php:
```diff
@@ -171,6 +171,30 @@ public function test_generate_cache_key_placeholder() {
+	 * @covers WP_Query::generate_cache_key
+	 * @dataProvider data_query_cache_duplicate
+	 */
+	public function test_generate_cache_key_normalize( $query_vars1, $query_vars2 ) {
```
Can you please add a ticket number to this test?

joemcgill reviewed (changes requested):
I think this is looking pretty good. A couple of nitpicks, but I really like this approach to handling this by normalizing the cache key generation, rather than modifying `get_posts()`. Nice!

jonnynews committed:
Sort post types for cache key generation

The code has been updated to sort post types before the cache key generation to ensure consistency. Also, a default orderby value of 'date' has been added for the same purpose. Changes have been tested by adding a new test case to verify the query cache for differently ordered post types.

jonnynews committed:
Fix lint error

drhsrushb commented:
o address this, you may need to ensure that the post_type variable is properly set in the sticky subquery, especially on the initial run. Double-checking how the post_type variable is initialized and passed in the query could help resolve the issue. Additionally, reviewing caching mechanisms and how they handle queries with different keys might provide further insights into how to prevent duplicate queries in this scenario.
{{{
#!html
<a href="https://bytevarsity.ca/"></a><a href="http://ukvibesonline.co.uk/"></a><a href="https://dcanews.ca/"></a><a href="https://canadahomesnews.ca/"></a><a href="https://blesa.ca/"></a><a href="https://vixrim.ca/"></a><a href="https://totimescanews.ca/"></a><a href="http://canadianews.ca/"></a><a href="http://newsag.ca/"></a><a href="http://ottawas.ca/"></a><a href="http://canadavista.ca/"></a><a href="https://cryptowavezen.com/"></a><a href="https://bitcryptomining.net/"></a><a href="https://cryptocoinsupdates.com/"></a><a href="https://forexcryptonews.net/"></a><a href="https://bitforexvista.com/"></a><a href="https://stocktradalert.com/"></a><a href="https://tradingsentiment.org/"></a><a href="https://forexinvester.com/"></a><a href="https://stocksnotify.com/"></a><a href="https://stocktrendsalert.com/"></a><a href="https://topforexebroker.com/"></a><a href="https://forexah.com/"></a><a href="https://forexinsides.com/"></a><a href="https://stocksrush.com/"></a>
}}}
If you're comfortable sharing more details or code snippets, I'd be happy to provide more specific suggestions or assistance!

joemcgill commented on src/wp-includes/class-wp-query.php:
```diff
@@ -2542,6 +2542,8 @@ public function get_posts() {
 			}
 		} elseif ( ! empty( $post_type ) && is_array( $post_type ) ) {
+			// Sort post types to ensure same cache key generation.
+			sort( $post_type );
```
I assume this needs to happen here as well as sorting the `post_type` args in `generate_cache_key()` in order to normalize the SQL as well?

jonnynews replied:
>I assume this needs to happen here as well as sorting the post_type args in generate_cache_key() in order to normalize the SQL as well?

Correct. A query for ```post_type IN ( 'page', 'post' )``` would get the same results as ```post_type IN ( 'post', 'page' )```

peterwilsoncc replied:
It's outside the diff by in the `any` block above this one:

```
$in_search_post_types = get_post_types( array( 'exclude_from_search' => false ) );
sort( $in_search_post_types );
```

As with this change, it will normalize the SQL

joemcgill reviewed (approved):
This is looking good to me.

peterwilsoncc commented on src/wp-includes/class-wp-query.php:
```diff
@@ -4856,6 +4858,24 @@ protected function generate_cache_key( array $args, $sql ) {
+				$args['post_type'] = 'post';
+			}
+		}
+		$args['post_type'] = (array) $args['post_type'];
```
I think this will cause problems with `any` and `all` post types: 

https://github.com/WordPress/wordpress-develop/blob/3b9aefb06cae1f34ed19b08caf3b3ed32df04b73/src/wp-includes/class-wp-query.php#L2536-L2544

Rajinsharwar replied:
I agree with @peterwilsoncc. Not sure, but maybe we can do add additional checks for 'any', and use something like below:

`$args['post_type'] = get_post_types( array( 'exclude_from_search' => false ) ) ?: '';`

spacedmonkey replied:
Good catch @peterwilsoncc. I put in a fix here and added a unit tests. https://github.com/WordPress/wordpress-develop/pull/6393/commits/533509b7fa8465bc80eb1109144e1994c56f6a42. 

As far as I know, post type all, is not a valid argument and is not documtened anywhere, This is a case that I haven't written for. Can you give me examples in the wild where this is used and how it would break?

Rajinsharwar replied:
I guess what is meant by all could be maybe this check if it's an array.

`! empty( $post_type ) && is_array( $post_type )`

spacedmonkey replied:
> I guess what is meant by all could be maybe this check if it's an array.
> 
> `! empty( $post_type ) && is_array( $post_type )`

Why is this check needed?

peterwilsoncc replied:
> As far as I know, post type all, is not a valid argument and is not documtened anywhere, This is a case that I haven't written for. Can you give me examples in the wild where this is used and how it would break?

That's because I was thinking of `post_status` (also undocumented but used in a few examples on .org) -- sorry about that.

jonnynews committed:
Update 'any' post type in WP_Query

The WP_Query has been updated to correctly handle 'any' as a valid post type argument. This includes a change where 'any' will now return all post types excluded from search. New tests have also been introduced to ensure cache keys behave as expected when post types are registered, unregistered, and clarified under 'any' classification.

jonnynews committed:
Add new test case to cacheResults.php

A new test scenario 'any and post types' has been added to the `cacheResults.php` testing module. This change aims to improve test coverage by checking query variables against different types of posts such as 'post', 'page', 'attachment' under the 'any' post type.

mukeshpanchal27 commented on tests/phpunit/tests/query/cacheResults.php:
```diff
@@ -171,6 +171,27 @@ public function test_generate_cache_key_placeholder() {
+	 * @covers WP_Query::generate_cache_key
+	 * @dataProvider data_query_cache_duplicate
+	 * @ticket 59442
+	 */
```
```suggestion
	/**
	 * @ticket 59442
	 *
	 * @covers WP_Query::generate_cache_key
	 *
	 * @dataProvider data_query_cache_duplicate
	 */
```

mukeshpanchal27 commented on tests/phpunit/tests/query/cacheResults.php:
```diff
@@ -216,6 +237,53 @@ public function test_query_cache( $args ) {
 		}
 	}
 
+	public function data_query_cache_duplicate() {
```
```suggestion
	/**
	 * Data provider for test_generate_cache_key_normalize().
	 *
	 * @return array[]
	 */
	 public function data_query_cache_duplicate() {
```

mukeshpanchal27 reviewed (commented):
Thanks for the PR! I left some nit-picks. Peter's feedback needs to be addressed before merging.

jonnynews committed:
Apply suggestions from code review

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

Rajinsharwar commented on tests/phpunit/tests/query/cacheResults.php:
```diff
@@ -216,6 +268,66 @@ public function test_query_cache( $args ) {
+	 *
+	 * @return array[]
+	 */
+	 public function data_query_cache_duplicate() {
```
```suggestion
    public function data_query_cache_duplicate() {
```

mukeshpanchal27 replied:
Only one space needs to remove.

mukeshpanchal27 commented on tests/phpunit/tests/query/cacheResults.php:
```diff
@@ -216,6 +268,66 @@ public function test_query_cache( $args ) {
 		}
 	}
 
+	/**
+	 * Data provider for test_generate_cache_key_normalize().
+	 *
+	 * @return array[]
+	 */
+	 public function data_query_cache_duplicate() {
```
```suggestion
	/**
	 * Data provider for test_generate_cache_key_normalize().
	 *
	 * @return array[]
	 */
	public function data_query_cache_duplicate() {
```
PHPCS

spacedmonkey committed:
Apply suggestions from code review

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

peterwilsoncc commented on tests/phpunit/tests/query/cacheResults.php:
```diff
@@ -171,6 +171,58 @@ public function test_generate_cache_key_placeholder() {
+		$reflection->setAccessible( true );
+
+		$cache_key_1 = $reflection->invoke( $query1, $query_vars1, $request1 );
+		$cache_key_2 = $reflection->invoke( $query1, $query_vars2, $request1 );
```
It's not yet defined but shouldn't this be using the request value from the second query.

jonnynews replied:
I tried to do this in a early versions of the patch. It just made the tests harder to follow. I think this is fine. We can improve the tests later.

peterwilsoncc reviewed (commented):
A couple of notes but looks good generally.

spacedmonkey committed:
Merge branch 'trunk' into fix/cache-key-generation

spacedmonkey committed:
Sort post types for consistent cache key generation

An update was made to the WP_Query class to ensure that when multiple post types are in use, they are sorted to generate a consistent cache key. This helps avoid potential issues with cache key mismatch and optimizes the performance of WordPress' cache system.

spacedmonkey commented:
Thanks for the feedback @peterwilsoncc . Actioned it here https://github.com/WordPress/wordpress-develop/pull/6393/commits/55026bcad3380ec1ee30e95379ba47be3c84994a

peterwilsoncc commented:
@spacedmonkey Sorry mate, I was thinking this block here. But it was good to catch the other spot.

https://github.com/WordPress/wordpress-develop/blob/55026bcad3380ec1ee30e95379ba47be3c84994a/src/wp-includes/class-wp-query.php#L2539-L2547

spacedmonkey committed:
Sort queried post types and status for consistent cache key generation

In the class-wp-query.php file, the queried post types are now sorted to ensure consistency in cache key generation. Similarly, post statuses are also sorted for the same reason. This change also impacted the related unit tests in cacheResults.php, which are updated to reflect this new sorting logic.

jonnynews commented:
> @spacedmonkey Sorry mate, I was thinking this block here. But it was good to catch the other spot.
> 
> https://github.com/WordPress/wordpress-develop/blob/55026bcad3380ec1ee30e95379ba47be3c84994a/src/wp-includes/class-wp-query.php#L2539-L2547

Good catch @peterwilsoncc. I have actioned your feedback found some more cases where keys would not match. 
- If post status is sent in a different order. 
- If post status is string and array. 

I also improved the tests to use second query object and use the second query objects request. This highlighted a weird bug, where spaces in the query resulted in none match sql, meaning the keys do not match. 

Tests are now passing.

peterwilsoncc commented on tests/phpunit/tests/query/cacheResults.php:
```diff
@@ -171,6 +171,63 @@ public function test_generate_cache_key_placeholder() {
+		unregister_post_type( 'wptests_pt' );
+		$cache_key_2 = $reflection->invoke( $query1, $query_vars, $request1 );
+
+		$this->assertNotSame( $cache_key_1, $cache_key_2, 'Cache key should differs when after unregister post type.' );
```
```suggestion
		$this->assertNotSame( $cache_key_1, $cache_key_2, 'Cache key should differ after unregistering post type.' );
```

peterwilsoncc commented on src/wp-includes/class-wp-query.php:
```diff
@@ -3107,11 +3112,11 @@ public function get_posts() {
 		// Beginning of the string is on a new line to prevent leading whitespace. See https://core.trac.wordpress.org/ticket/56841.
```
```suggestion
		/*
		 * Beginning of the string is on a new line to prevent leading whitespace. 
		 *
		 * The additional indentation of subsequent lines is to ensure the SQL
		 * queries are identical to those generated when splitting queries. This
		 * improves caching of the query by ensuring the same cache key is 
		 * generated for the same database queries functionally.
		 *
		 * See https://core.trac.wordpress.org/ticket/56841.
		 * See https://github.com/WordPress/wordpress-develop/pull/6393#issuecomment-2088217429
		 */
```

peterwilsoncc reviewed (approved):
This looks good to me, just a couple of notes to fix a typo and improve a comment.

Pre-approving as neither of the suggestions related to code.

jonnynews committed:
Apply suggestions from code review

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

spacedmonkey committed:
Merge branch 'trunk' into fix/cache-key-generation

spacedmonkey committed:
Fix lint errors.

spacedmonkey committed:
Merge branch 'trunk' into fix/cache-key-generation

spacedmonkey commented:
Committed in https://core.trac.wordpress.org/changeset/58122
